ext/uri/php_uriparser.c

```diff
@@ -293,51 +292,68 @@ static uriparser_uris_t *uriparser_create_uris(void)
 	return uriparser_uris;
 }
 
-static void throw_invalid_uri_exception(void)
+void *uriparser_parse_uri_ex(const zend_string *uri_str, const uriparser_uris_t *uriparser_base_urls, bool silent)
 {
-	zend_throw_exception(uri_invalid_uri_exception_ce, "The specified URI is malformed", 0);
-}
+	UriUriA uri = {0};
 
-#define PARSE_URI(dest_uri, uri_str, uriparser_uris, silent) \
-	do { \
-		if (ZSTR_LEN(uri_str) == 0 || \
-			uriParseSingleUriExMmA(dest_uri, ZSTR_VAL(uri_str), ZSTR_VAL(uri_str) + ZSTR_LEN(uri_str), NULL, mm) != URI_SUCCESS \
-		) { \
-			efree(uriparser_uris); \
-			if (!silent) { \
-				throw_invalid_uri_exception(); \
-			} \
-			return NULL; \
-		} \
-	} while (0)
+	/* Empty URIs are always invalid. */
+	if (ZSTR_LEN(uri_str) == 0) {
+		if (!silent) {
+			zend_throw_exception(uri_invalid_uri_exception_ce, "The specified URI must not be empty", 0);
```

Comment:
I wonder if empty inputs should be a ValueError for both Rfc3986 and WhatWg, since this is something that the user can reasonably check before trying to parse.

Reply:
> Although one exception type is likely more convenient.

ValueErrors are not intended to be caught, so this specifically would be something “you should've checked this before and it's reasonable to expect this from you and the fact that you didn't likely indicates a bug elsewhere in your application”.